# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: security/sandbox/chromium/base/tuple.h
# Commit: 370c40e49230
# Full Hash: 370c40e4923073a87845b57c1f4851bacd9663da
# Author: Bob Owen <bobowencode@gmail.com>
# Date: 2019-06-12 21:42:35
# Regressor Bug: 1552160
# File Overlap Count: 3
# Description:
#   Bug 1552160 Part 1: Roll-up of chromium sandbox update and mozilla patches to get a running browser. r=jld,aklotz,tjr,bobowen
#   
#   This updates security/sandbox/chromium/ files to chromium commit 84108231f6e6e0772fb9a4643679ce76aa771e67.
#   
#   Existing and new patches applied from security/sandbox/chromium-shim/patches/with_update/ to give a compiling and mostly working browser.
# ==============================================================================

diff -r bd5ecd72f76c -r 370c40e49230 security/sandbox/chromium/base/tuple.h
--- a/security/sandbox/chromium/base/tuple.h	Wed Jun 12 21:55:59 2019 +1200
+++ b/security/sandbox/chromium/base/tuple.h	Wed Jun 12 11:10:48 2019 +0100
@@ -27,37 +27,12 @@
 
 #include <stddef.h>
 #include <tuple>
+#include <utility>
 
 #include "build/build_config.h"
 
 namespace base {
 
-// Index sequences
-//
-// Minimal clone of the similarly-named C++14 functionality.
-
-template <size_t...>
-struct IndexSequence {};
-
-template <size_t... Ns>
-struct MakeIndexSequenceImpl;
-
-template <size_t... Ns>
-struct MakeIndexSequenceImpl<0, Ns...> {
-  using Type = IndexSequence<Ns...>;
-};
-
-template <size_t N, size_t... Ns>
-struct MakeIndexSequenceImpl<N, Ns...>
-    : MakeIndexSequenceImpl<N - 1, N - 1, Ns...> {};
-
-template <size_t N>
-using MakeIndexSequence = typename MakeIndexSequenceImpl<N>::Type;
-
-template <typename T>
-using MakeIndexSequenceForTuple = MakeIndexSequence<
-    std::tuple_size<typename std::decay<T>::type>::value>;
-
 // Dispatchers ----------------------------------------------------------------
 //
 // Helper functions that call the given method on an object, with the unpacked
@@ -73,7 +48,7 @@
 inline void DispatchToMethodImpl(const ObjT& obj,
                                  Method method,
                                  Tuple&& args,
-                                 IndexSequence<Ns...>) {
+                                 std::index_sequence<Ns...>) {
   (obj->*method)(std::get<Ns>(std::forward<Tuple>(args))...);
 }
 
@@ -81,8 +56,9 @@
 inline void DispatchToMethod(const ObjT& obj,
                              Method method,
                              Tuple&& args) {
+  constexpr size_t size = std::tuple_size<std::decay_t<Tuple>>::value;
   DispatchToMethodImpl(obj, method, std::forward<Tuple>(args),
-                       MakeIndexSequenceForTuple<Tuple>());
+                       std::make_index_sequence<size>());
 }
 
 // Static Dispatchers with no out params.
@@ -90,14 +66,15 @@
 template <typename Function, typename Tuple, size_t... Ns>
 inline void DispatchToFunctionImpl(Function function,
                                    Tuple&& args,
-                                   IndexSequence<Ns...>) {
+                                   std::index_sequence<Ns...>) {
   (*function)(std::get<Ns>(std::forward<Tuple>(args))...);
 }
 
 template <typename Function, typename Tuple>
 inline void DispatchToFunction(Function function, Tuple&& args) {
+  constexpr size_t size = std::tuple_size<std::decay_t<Tuple>>::value;
   DispatchToFunctionImpl(function, std::forward<Tuple>(args),
-                         MakeIndexSequenceForTuple<Tuple>());
+                         std::make_index_sequence<size>());
 }
 
 // Dispatchers with out parameters.
@@ -112,8 +89,8 @@
                                  Method method,
                                  InTuple&& in,
                                  OutTuple* out,
-                                 IndexSequence<InNs...>,
-                                 IndexSequence<OutNs...>) {
+                                 std::index_sequence<InNs...>,
+                                 std::index_sequence<OutNs...>) {
   (obj->*method)(std::get<InNs>(std::forward<InTuple>(in))...,
                  &std::get<OutNs>(*out)...);
 }
@@ -123,9 +100,11 @@
                              Method method,
                              InTuple&& in,
                              OutTuple* out) {
+  constexpr size_t in_size = std::tuple_size<std::decay_t<InTuple>>::value;
+  constexpr size_t out_size = std::tuple_size<OutTuple>::value;
   DispatchToMethodImpl(obj, method, std::forward<InTuple>(in), out,
-                       MakeIndexSequenceForTuple<InTuple>(),
-                       MakeIndexSequenceForTuple<OutTuple>());
+                       std::make_index_sequence<in_size>(),
+                       std::make_index_sequence<out_size>());
 }
 
 }  // namespace base